docs/pencil_options.md
# Pencil Options
The pencil configuration files is written in YAML. When pencil starts up, it
searches pencil.yml and loads it (use -f to specify a particular file). See the
main README.md for more details.

## General Configuration

* :graphite_url [no default]

  The url of your graphite instance.

* :default_url_opts [optional, no default]

  A map of default graph options, like width and height. See the
  [DSL](https://github.com/ripienaar/graphite-graph-dsl/wiki) for specific
  options. A sample configuration might would look like:

    :default_url_opts:
      :width: 1400
      :height: 400
      :fontsize: 15

* :refresh_rate [default 60]

  How often to refresh graph images, in seconds.

  In the future this option will be overridden under the :default_views: key
  per view, this value being the default.

* :host_sort [default "sensible"]

  One of "builtin", "numeric", or "sensible".

  Set to "builtin" to sort using ruby's builtin String sort.

  Set to "numeric" to sort hosts numerically (i.e. match secondarily on the
  first \d+).

  Set to "sensible" if you want to sort like this:

  http://www.bofh.org.uk/2007/12/16/comprehensible-sorting-in-ruby

* :metric_format [default "%m.%c.%h"]

  The format your graphite metrics are stored in. Pencil needs metrics to be
  composed of two or three distinct pieces, concatenated in some regular
  fashion. The format strings are:

  * %m metric
  * %c cluster
  * %h host

  The cluster "%c" portion is optional.

  Currently supported metric types are those that are qualified by at most a
  cluster and a short hostname, such as

  [METRIC].colo1.db1 # corresponding to db1.colo1.[rest of FQDN]
  
  or, with no cluster:

  [METRIC].db1 # corresponding to db1.[rest of FQDN]

  Example metric and corresponding formats:

  system.load.hostname                          # %m.%h
  system.load.dc.hostname                       # %m.%c.%h
  dc.hostname.system.load                       # %c.%h.%m
  hostname.dc.system.load                       # %h.%c.%m
  prefix_for_all_metrics.system.load.hostname   # prefix_for_all_metrics.%m.%h

  Pencil will recognize all these graphite metrics as "system.load" for the
  host "hostname", and, where applicable, cluster "dc".

* :port [default 9292]

   Port to bind to.

* :templates_dir [no default]

   Directory where graph and dashboard configuration is stored. Resolved
   relative to the location of the config file.

* :webapp [default false]

   Enable Open Web App support. See [here](./webapp.md) for details.

* :default_views:

  Pencil has two modes: live views and calendar views. For live monitoring of
  graphite data live views are used. This key species the default views
  available (though manually specifying "from=RELATIVE" in the URL works too).

  Default value:

    :default_views:
      - -1h: "one hour view"
      - default:
         -8h: "eight hour view"
      - -1day: "One day view"

 Each entry in this list should look like one of the following:

 TIMESPEC
 TIMESPEC: LABEL
 default: TIMESPEC
 default: TIMESPEC: LABEL

 TIMESPEC is a graphite at-style relative time from
 "now". [See the documentation](https://graphite.readthedocs.org/en/latest/render_api.html#from-until)
 for details.

 *NOTE* acceptable values for this key will expand in a later release to
 include arbitrary graphite-supported timespecs (not just references), an
 additional supported "until" value, and a "refresh" value as well. It will
 remain backwards compatible.


lib/pencil/config.rb
require 'logger'
require 'pathname'
require 'thread'
require 'find'
require 'optparse'
require 'json'
require 'open-uri'
require 'yaml'
require 'set'

require 'pencil/models'
require 'pencil/attime'

module Pencil
  class Config
    include Pencil::Models

    attr_reader :dashboards
    attr_reader :graphs
    attr_reader :hosts
    attr_reader :clusters
    attr_reader :clustermap
    attr_reader :config_file
    attr_reader :version
    attr_reader :port
    attr_reader :argv

    def initialize
      @port = nil
      @config_file = File.expand_path './pencil.yml'
      @recursive = false
      @logger = Logger.new(STDOUT) # fixme use the sinatra logger
      @argv = ARGV.clone

      optparse = OptionParser.new do |o|
        o.on('-d', '--config-dir DIR',
             'deprecated: specify a config file instead with -f') do |arg|
          abort '-d deprecated: specify a config file instead with -f'
        end
        o.on('-f', '--config-file FILE',
             'location of the config file (default ./pencil.yml)') do |arg|
          @config_file = File.expand_path arg
        end
        o.on('-p', '--port PORT', 'port to bind to (default 9292)') do |arg|
          @port = arg.to_i
        end
        o.on('-V', '--version', 'print version string and exit') do
          puts Pencil::VERSION
          exit 0
        end
      end

      optparse.parse(@argv)
      stage_load
      @port ||= 9292
      reload!
    end

    def [] (key)
      @config[key]
    end

    # for views
    class << self
      attr_accessor :klass
    end
    def klass
      self.class.klass ||= Struct.new(:from, :offset, :label, :is_default)
    end

    # fixme handle exceptions in ATTime code
    def gen_view (h)
      if h.is_a? String
        klass.new(h, ATTime.parseTimeOffset(h), nil)
      elsif h.keys.first.to_s == 'default'
        if h.values.first.is_a? String
          klass.new(h.values.first, ATTime.parseTimeOffset(h.values.first), nil, true)
        else
          klass.new(h.values.first.keys.first,
                    ATTime.parseTimeOffset(h.values.first.keys.first),
                    h.values.first.values.first, true)
        end
      else
        klass.new(h.keys.first, ATTime.parseTimeOffset(h.keys.first), h.values.first)
      end
    end

    # fixme re-implement all the config reload stuff
    def stage_load
      defaults = {
        :host_sort => 'sensible',
        :metric_format => '%m.%c.%h',
        :refresh_rate => 60,
        :default_views =>
        [{'-1h' => 'one hour view'},
         {'default' => {'-8h' => 'eight hour view'}},
         {'-1day' => 'one day view'}]
      }
      unless File.readable?(@config_file) && File.file?(@config_file)
        abort "config file #{@config_file} not found or not readable (-f)"
      end

      yaml = YAML::load_file @config_file
      abort "couldn't parse #{@config_file} as YAML" unless yaml
      @_config = defaults.merge(yaml)

      # do some sanity checking of other configuration parameters
      [:graphite_url, :templates_dir].each do |c|
        if not yaml[c]
          abort "Missing config name ':#{c.to_s}'"
        end
      end
      PencilGraph.graphite_url = yaml[:graphite_url]
      PencilGraph.metric_format = yaml[:metric_format]

      @templates_dir = yaml[:templates_dir]

      if Pathname.new(@templates_dir).relative?
        @templates_dir = File.expand_path(@templates_dir, File.dirname(@config_file))
      end

      @templates_dir = @templates_dir[0..-2] if @templates_dir[-1..-1] == '/'

      unless File.readable?(@templates_dir) && File.directory?(@templates_dir)
        abort "templates directory #{@templates_dir} not found or not readable"
      end

      @port ||= @_config[:port] if @_config[:port]

      if @_config[:webapp]
        @_config[:webapp] = {} if @_config[:webapp] == true
        unless @_config[:webapp][:manifest]
          @logger.info "no :manifest key for webapp, using default"
          @_config[:webapp][:manifest] = {
            "name" => "Pencil",
            "description" => "Graphite Dashboard Frontend",
            "launch_path" => "/",
            "icons" => {
              "16" => "/favicon.ico"
            },
            "developer" => {
              "name" => "whd@mozilla.com",
              "url" => "https://github.com/fetep/pencil"
            },
            "default_locale" => "en"
          }
        end
      end

      @_config[:default_views].map! {|h| gen_view(h)}

      default = @_config[:default_views].select {|x| x.is_default}
      if default.size > 1
        @logger.warn "multiple default timeslices, using first: #{@_config[:default_views].first.from}"
        default[1..-1].each {|x| x.is_default = false}
      elsif default.size == 0
        @logger.warn "no default timeslice, using first: #{@_config[:default_views].first.from}"
        @_config[:default_views].first.is_default = true
      end

      # fixme warn on duplicate names, error checking
      # load graph definitions first
      @_graphs = {}
      @_dashboards = {}
      Dir.glob("#{@templates_dir}/**/*.graph").each do |f|
        g = PencilGraph.new(f, yaml[:default_url_opts])
        @_graphs[g.name] = g
      end
      # fixme special case ignore @config_file matches
      Dir.glob("#{@templates_dir}/**/*.y{a,}ml").map do |f|
        d = Dashboard.new(f, @templates_dir)
        @_dashboards[d.name] = d
      end

      if yaml[:metric_format] !~ /%m/
        abort 'missing metric (%m) in :metric_format'
      elsif yaml[:metric_format] !~ /%h/
        abort 'missing host (%h) in :metric_format'
      end

      @_hosts = {}
      clusters = SortedSet.new
      clustermap = {} # cluster name => cluster

      # fixme not reload safe
      Host.sort_method = @_config[:host_sort]
      @_graphs.each do |name, graph|
        hosts = graph.discover_hosts
        hosts.each do |hname, cluster|
          name = Host.get_name(hname, cluster)
          @_hosts[name] ||= Host.new(hname, cluster)
          @_hosts[name].graphs << graph.name
          clusters << cluster if cluster
          clustermap[cluster||:global] ||= SortedSet.new
          clustermap[cluster||:global] << @_hosts[name]
        end
      end

      noassoc = {} # cluster => hosts
      @_hosts.each do |name, h|
        h.graphs.sort!
        assoc = false
        @_dashboards.each do |dname, d|
          d.graphs.each do |g, hash|
            hash['hosts'].each do |w|
              if h.match(w)
                d.clusters << h.cluster if h.cluster
                d.assoc[g][w] << h
                assoc = true
              end
            end
          end
        end
        unless assoc
          @logger.info "#{h.name} not associated with a dashboard"
          noassoc[h.cluster||:global] ||= SortedSet.new
          noassoc[h.cluster||:global] << h
        end
      end

      @_clusters = []
      if clusters.size == 0
        @_clusters <<
          Cluster.new(nil, clustermap[:global], noassoc[:global], true)
      else
        clusters.each do |c|
          @_clusters << Cluster.new(c, clustermap[c], noassoc[c])
        end
      end
      @_clusters.sort!
    end

    def load_verify_stage
      begin
        # save class variables
        @logger.info 'staging load'
        [PencilGraph, Dashboard, Host].each(&:save)
        [PencilGraph, Dashboard].each(&:clear)
        stage_load
        @logger.info 'staging load succeeded, reloading'
        reload!
      rescue Exception => err
        [PencilGraph, Dashboard, Host].map(&:restore)
        @logger.error "error reloading configuration:\n#{err}"
        @logger.warn "staging load failed, using old configuration #{@version}"
        return false
      end
      return true
    end

    def reload!
      @dashboards, @graphs = @_dashboards, @_graphs
      @hosts, @clusters = @_hosts, @_clusters
      @config = @_config
      @clustermap = @_clustermap
      @_clustermap = @_config = @_dashboards = nil
      @_graphs = @_hosts = @_clusters = nil
      @version = Time.now.to_i
    end

  end # Pencil::Config
end # Pencil
